staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -217,54 +297,157 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+// LoadApiServiceSpec loads OpenAPI spec for the given API Service and then updates aggregator's spec.  It is thread safe.
+func (s *openAPIAggregator) LoadApiServiceSpec(handler http.Handler, apiService *apiregistration.APIService) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 
 	// Ignore local services
 	if apiService.Spec.Service == nil {
 		return nil
 	}
 
-	openApiSpec, err := s.downloadOpenAPISpec(handler)
+	openApiSpec, etag, _, err := s.downloadOpenAPISpec(handler, "")
 	if err != nil {
 		return err
 	}
+	if openApiSpec == nil {
+		// API service does not provide an OpenAPI spec.
+		s.removeApiServiceSpec(apiService.Name)
+		return nil
+	}
 	aggregator.FilterSpecByPaths(openApiSpec, []string{"/apis/" + apiService.Spec.Group + "/"})
 
+	_, existingService := s.openAPISpecs[apiService.Name]
 	s.openAPISpecs[apiService.Name] = &openAPISpecInfo{
 		apiService: *apiService,
 		spec:       openApiSpec,
+		handler:    handler,
+		etag:       etag,
 	}
 
 	err = s.updateOpenAPISpec()
 	if err != nil {
 		delete(s.openAPISpecs, apiService.Name)
 		return err
 	}
+	if !existingService {
+		s.openAPIAggregationController.enqueue(apiService.Name)
+	}
+
+	return nil
+}
+
+// RemoveApiServiceSpec removes an api service from OpenAPI aggregation. It is thread safe.
+func (s *openAPIAggregator) RemoveApiServiceSpec(apiServiceName string) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	return s.removeApiServiceSpec(apiServiceName)
+}
+
+// removeApiServiceSpec is the internal version of RemoveApiServiceSpec that does not hold any lock and safe to call
+// from a method that holds the lock.
+func (s *openAPIAggregator) removeApiServiceSpec(apiServiceName string) error {
+	if _, existingService := s.openAPISpecs[apiServiceName]; !existingService {
+		return nil
+	}
+	oldSpecs := s.openAPISpecs
+	delete(s.openAPISpecs, apiServiceName)
+	err := s.updateOpenAPISpec()
+	if err != nil {
+		s.openAPISpecs = oldSpecs
+		return err
+	}
 	return nil
 }
+
+// UpdateApiServiceSpec updates and aggregation OpenAPI spec for given service name. It is thread safe.
+func (s *openAPIAggregator) UpdateApiServiceSpec(apiServiceName string) (shouldBeRateLimited, deleted bool, err error) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	specInfo, existingService := s.openAPISpecs[apiServiceName]
+	if !existingService {
+		return false, true, nil
+	}
+	// If this is a local spec with no handler, that means it is a static spec.
+	if specInfo.handler == nil {
+		return false, false, nil
+	}
+	spec, etag, httpStatus, err := s.downloadOpenAPISpec(specInfo.handler, specInfo.etag)
+	switch {
+	case err != nil:
+		return true, false, err
+	case httpStatus == http.StatusNotModified:
+		return false, false, nil
+	case httpStatus == http.StatusNotFound || spec == nil:
+		return true, false, nil
+	}
+
+	oldSpec := specInfo.spec
+	specInfo.spec = spec
+	if err := s.updateOpenAPISpec(); err != nil {
+		specInfo.spec = oldSpec
```

Comment:
No, `updateOpenAPISpec` does not change specInfo. What we are doing here is we set the spec of API Service to the new one, try to aggregate it, if we failed to aggregated (possible because of merge problems or deserialization problems, etc.) that means we still serving the old spec for this APIService as we didn't change aggregated spec (that's the logic of don't change anything on error for `updateOpenAPISpec`). So we need to restore old spec in specInfo to have the right input for aggregation. If we don't do this, possibly, all later aggregations will fail as we have a bad input in specInfo for this APIService. Does that make sense?